fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreTest.java

```diff
@@ -3272,4 +3272,68 @@ public IndexState needRebuildIndex(Index index, long recordCount, boolean indexO
             assertTrue(recordStore.getRecordStoreState().getStoreHeader().getOmitUnsplitRecordSuffix());
         }
     }
+
+    @Test
+    public void testIndexStateFormatVersionUpgrade() {
+        final RecordMetaDataBuilder builder = RecordMetaData.newBuilder().setRecords(TestNoIndexesProto.getDescriptor());
+        try (FDBRecordContext context = openContext()) {
+            recordStore = FDBRecordStore.newBuilder()
+                    .setContext(context)
+                    .setMetaDataProvider(builder)
+                    .setKeySpacePath(path)
+                    .setFormatVersion(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1)
+                    .create();
+            assertEquals(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1, recordStore.getFormatVersion());
+            // Save 200 records without any indexes.
+            for (int i = 0; i < 200; i++) {
+                TestNoIndexesProto.MySimpleRecord record = TestNoIndexesProto.MySimpleRecord.newBuilder()
+                        .setRecNo(i).setNumValue(i + 1000).build();
+                recordStore.saveRecord(record);
+            }
+            commit(context);
+        }
+
+        // Add a new index named "index" on rec_no.
+        final Index index = new Index("index", "rec_no");
+        index.setSubspaceKey(1);
+        builder.addIndex("MySimpleRecord", index);
+        try (FDBRecordContext context = openContext()) {
+            recordStore = FDBRecordStore.newBuilder()
+                    .setContext(context)
+                    .setMetaDataProvider(builder)
+                    .setKeySpacePath(path)
+                    .setFormatVersion(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1)
+                    .createOrOpen();
+            assertEquals(FDBRecordStore.SAVE_INDEX_STATE_WITH_SUBSPACE_KEY_FORMAT_VERSION - 1, recordStore.getFormatVersion());
+            // Since there were 200 records already, the new index is write-only.
+            assertEquals(IndexState.WRITE_ONLY, recordStore.getRecordStoreState().getState(index.getName()));
+            assertEquals(1L, recordStore.getRecordMetaData().getIndex("index").getSubspaceKey());
+            context.commit();
```

Comment:
It would be good to validate here that the old state location is written to in a format that we expect older format versioned things to be be able to read.

Reply:
sure.